File: fs/before_checksum/460__lfu-cache/cpp/lfu-cache_benchmark_v2.cpp
#include <iostream>
#include <unordered_map>
#include <list>
#include <climits>
#include <utility>
#include <algorithm>
#include <vector>

using namespace std;

// Time:  O(1), per operation
// Space: O(k), k is the capacity of cache

// using OrderedDict
class LFUCache {
public:
    LFUCache(int capacity) : capacity_(capacity), size_(0), min_freq_(INT_MAX) {}

    int get(int key) {
        if (key_to_freq_.find(key) == key_to_freq_.end()) {
            return -1;
        }
        int freq = key_to_freq_[key];
        int value = freq_to_nodes_[freq].get_value(key);
        update(key, value);
        return value;
    }

    void put(int key, int value) {
        if (capacity_ <= 0) {
            return;
        }
        if (key_to_freq_.find(key) == key_to_freq_.end() && size_ == capacity_) {
            auto& od = freq_to_nodes_[min_freq_];
            int evict_key = od.pop_front_key();
            key_to_freq_.erase(evict_key);
            if (od.empty()) {
                freq_to_nodes_.erase(min_freq_);
            }
            --size_;
        }
        update(key, value);
    }

private:
    struct OrderedDict {
        list<int> order;
        unordered_map<int, pair<list<int>::iterator, int>> m; // key -> (iterator in order, value)

        void set(int key, int value) {
            auto it = m.find(key);
            if (it != m.end()) {
                it->second.second = value;
            } else {
                order.push_back(key);
                auto iter = prev(order.end());
                m.emplace(key, make_pair(iter, value));
            }
        }

        void erase(int key) {
            auto it = m.find(key);
            if (it == m.end()) return;
            order.erase(it->second.first);
            m.erase(it);
        }

        bool empty() const {
            return m.empty();
        }

        int pop_front_key() {
            int k = order.front();
            order.pop_front();
            m.erase(k);
            return k;
        }

        int get_value(int key) const {
            auto it = m.find(key);
            return it->second.second;
        }
    };

    void update(int key, int value) {
        int freq = 0;
        auto it = key_to_freq_.find(key);
        if (it != key_to_freq_.end()) {
            freq = it->second;
            auto fit = freq_to_nodes_.find(freq);
            if (fit != freq_to_nodes_.end()) {
                fit->second.erase(key);
                if (fit->second.empty()) {
                    freq_to_nodes_.erase(fit);
                    if (min_freq_ == freq) {
                        ++min_freq_;
                    }
                }
            }
            --size_;
        }

        ++freq;
        min_freq_ = std::min(min_freq_, freq);
        key_to_freq_[key] = freq;
        freq_to_nodes_[freq].set(key, value);
        ++size_;
    }

    int capacity_;
    int size_;
    int min_freq_;
    unordered_map<int, OrderedDict> freq_to_nodes_;
    unordered_map<int, int> key_to_freq_;
};

struct Op {
    char type; // 'p' for put, 'g' for get
    int key;
    int value; // used only for put
};

struct Test {
    int capacity;
    vector<Op> ops;
};

int main() {
    vector<Test> tests;

    // 1
    tests.push_back({2, {
        {'p',1,1}, {'p',2,2}, {'g',1,0}, {'p',3,3}, {'g',2,0},
        {'g',3,0}, {'p',4,4}, {'g',1,0}, {'g',3,0}, {'g',4,0}
    }});

    // 2
    tests.push_back({1, {
        {'p',1,10}, {'g',1,0}, {'p',2,20}, {'g',1,0}, {'g',2,0},
        {'p',2,200}, {'g',2,0}
    }});

    // 3
    tests.push_back({3, {
        {'p',1,1}, {'p',2,2}, {'p',3,3}, {'g',1,0}, {'g',2,0},
        {'p',4,4}, {'g',3,0}, {'g',4,0}, {'g',1,0}, {'p',5,5},
        {'g',2,0}, {'g',5,0}
    }});

    // 4
    tests.push_back({3, {
        {'p',1,100}, {'p',2,200}, {'p',3,300}, {'g',1,0}, {'g',1,0},
        {'g',1,0}, {'g',2,0}, {'p',4,400}, {'g',3,0}, {'g',1,0}, {'g',4,0}
    }});

    // 5
    tests.push_back({2, {
        {'p',1,1}, {'p',2,2}, {'g',1,0}, {'p',1,10}, {'g',1,0},
        {'p',3,3}, {'g',2,0}, {'g',3,0}
    }});

    // 6
    tests.push_back({4, {
        {'p',1,1}, {'p',2,2}, {'p',3,3}, {'p',4,4}, {'g',1,0}, {'g',2,0},
        {'p',5,5}, {'g',3,0}, {'g',4,0}, {'p',6,6}, {'g',4,0}, {'g',5,0}, {'g',6,0}
    }});

    // 7
    tests.push_back({3, {
        {'p',1,10}, {'p',2,20}, {'p',3,30}, {'g',2,0}, {'p',4,40},
        {'g',1,0}, {'g',3,0}, {'g',2,0}
    }});

    // 8
    tests.push_back({5, {
        {'p',1,10}, {'p',2,20}, {'p',3,30}, {'p',4,40}, {'p',5,50},
        {'g',1,0}, {'g',1,0}, {'p',1,100}, {'g',1,0}, {'p',6,60},
        {'g',2,0}, {'g',6,0}
    }});

    // 9
    tests.push_back({2, {
        {'p',1,1}, {'g',1,0}, {'g',1,0}, {'g',1,0}, {'g',1,0}, {'g',1,0},
        {'p',2,2}, {'p',3,3}, {'g',1,0}, {'g',2,0}, {'g',3,0}
    }});

    // 10
    tests.push_back({3, {
        {'p',0,0}, {'p',1,1}, {'p',2,2}, {'g',0,0}, {'g',1,0},
        {'p',3,3}, {'g',2,0}, {'p',0,5}, {'g',0,0}
    }});

    volatile int sink = 0;
    const int iterations = 1;
    for (int iter = 0; iter < iterations; ++iter) {
        int checksum = 0;
        for (const auto& test : tests) {
            LFUCache cache(test.capacity);
            int local = 0;
            for (const auto& op : test.ops) {
                if (op.type == 'p') {
                    cache.put(op.key, op.value);
                } else if (op.type == 'g') {
                    local += cache.get(op.key);
                }
            }
            checksum += local;
        }
        sink += checksum;
        

    } std::cout << sink << std::endl;

    return 0;
}

File: fs/py_before_energy/460__lfu-cache/lfu-cache_benchmark_v2.py
# Time:  O(1), per operation
# Space: O(k), k is the capacity of cache

import collections


# using OrderedDict
class LFUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.__capa = capacity
        self.__size = 0
        self.__min_freq = float("inf")
        self.__freq_to_nodes = collections.defaultdict(collections.OrderedDict)
        self.__key_to_freq = {}

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.__key_to_freq:
            return -1
        value = self.__freq_to_nodes[self.__key_to_freq[key]][key]
        self.__update(key, value)
        return value

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if self.__capa <= 0:
            return

        if key not in self.__key_to_freq and self.__size == self.__capa:
            del self.__key_to_freq[self.__freq_to_nodes[self.__min_freq].popitem(last=False)[0]]
            if not self.__freq_to_nodes[self.__min_freq]:
                del self.__freq_to_nodes[self.__min_freq]
            self.__size -= 1
        self.__update(key, value)
        
    def __update(self, key, value):
        freq = 0
        if key in self.__key_to_freq:
            freq = self.__key_to_freq[key]
            del self.__freq_to_nodes[freq][key]
            if not self.__freq_to_nodes[freq]:
                del self.__freq_to_nodes[freq]
                if self.__min_freq == freq:
                    self.__min_freq += 1
            self.__size -= 1

        freq += 1
        self.__min_freq = min(self.__min_freq, freq)
        self.__key_to_freq[key] = freq
        self.__freq_to_nodes[freq][key] = value
        self.__size += 1


# Time:  O(1), per operation
# Space: O(k), k is the capacity of cache
import collections


class ListNode(object):
    def __init__(self, key, value, freq):
        self.key = key
        self.val = value
        self.freq = freq
        self.next = None
        self.prev = None


class LinkedList(object):
    def __init__(self):
        self.head = None
        self.tail = None

    def append(self, node):
        node.next, node.prev = None, None  # avoid dirty node
        if self.head is None:
            self.head = node
        else:
            self.tail.next = node
            node.prev = self.tail
        self.tail = node

    def delete(self, node):
        if node.prev:
            node.prev.next = node.next
        else:
            self.head = node.next
        if node.next:
            node.next.prev = node.prev
        else:
            self.tail = node.prev
        node.next, node.prev = None, None  # make node clean


class LFUCache2(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.__capa = capacity
        self.__size = 0
        self.__min_freq = float("inf")
        self.__freq_to_nodes = collections.defaultdict(LinkedList)
        self.__key_to_node = {}

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.__key_to_node:
            return -1
        value = self.__key_to_node[key].val
        self.__update(key, value)
        return value

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if self.__capa <= 0:
            return

        if key not in self.__key_to_node and self.__size == self.__capa:
            del self.__key_to_node[self.__freq_to_nodes[self.__min_freq].head.key]
            self.__freq_to_nodes[self.__min_freq].delete(self.__freq_to_nodes[self.__min_freq].head)
            if not self.__freq_to_nodes[self.__min_freq].head:
                del self.__freq_to_nodes[self.__min_freq]
            self.__size -= 1
        self.__update(key, value)
        
    def __update(self, key, value):
        freq = 0
        if key in self.__key_to_node:
            old_node = self.__key_to_node[key]
            freq = old_node.freq
            self.__freq_to_nodes[freq].delete(old_node)
            if not self.__freq_to_nodes[freq].head:
                del self.__freq_to_nodes[freq]
                if self.__min_freq == freq:
                    self.__min_freq += 1
            self.__size -= 1

        freq += 1
        self.__min_freq = min(self.__min_freq, freq)
        self.__key_to_node[key] = ListNode(key, value, freq)
        self.__freq_to_nodes[freq].append(self.__key_to_node[key])
        self.__size += 1


if __name__ == "__main__":
    tests = []

    # 1
    tests.append((2, [
        ('p',1,1), ('p',2,2), ('g',1,0), ('p',3,3), ('g',2,0),
        ('g',3,0), ('p',4,4), ('g',1,0), ('g',3,0), ('g',4,0)
    ]))

    # 2
    tests.append((1, [
        ('p',1,10), ('g',1,0), ('p',2,20), ('g',1,0), ('g',2,0),
        ('p',2,200), ('g',2,0)
    ]))

    # 3
    tests.append((3, [
        ('p',1,1), ('p',2,2), ('p',3,3), ('g',1,0), ('g',2,0),
        ('p',4,4), ('g',3,0), ('g',4,0), ('g',1,0), ('p',5,5),
        ('g',2,0), ('g',5,0)
    ]))

    # 4
    tests.append((3, [
        ('p',1,100), ('p',2,200), ('p',3,300), ('g',1,0), ('g',1,0),
        ('g',1,0), ('g',2,0), ('p',4,400), ('g',3,0), ('g',1,0), ('g',4,0)
    ]))

    # 5
    tests.append((2, [
        ('p',1,1), ('p',2,2), ('g',1,0), ('p',1,10), ('g',1,0),
        ('p',3,3), ('g',2,0), ('g',3,0)
    ]))

    # 6
    tests.append((4, [
        ('p',1,1), ('p',2,2), ('p',3,3), ('p',4,4), ('g',1,0), ('g',2,0),
        ('p',5,5), ('g',3,0), ('g',4,0), ('p',6,6), ('g',4,0), ('g',5,0), ('g',6,0)
    ]))

    # 7
    tests.append((3, [
        ('p',1,10), ('p',2,20), ('p',3,30), ('g',2,0), ('p',4,40),
        ('g',1,0), ('g',3,0), ('g',2,0)
    ]))

    # 8
    tests.append((5, [
        ('p',1,10), ('p',2,20), ('p',3,30), ('p',4,40), ('p',5,50),
        ('g',1,0), ('g',1,0), ('p',1,100), ('g',1,0), ('p',6,60),
        ('g',2,0), ('g',6,0)
    ]))

    # 9
    tests.append((2, [
        ('p',1,1), ('g',1,0), ('g',1,0), ('g',1,0), ('g',1,0), ('g',1,0),
        ('p',2,2), ('p',3,3), ('g',1,0), ('g',2,0), ('g',3,0)
    ]))

    # 10
    tests.append((3, [
        ('p',0,0), ('p',1,1), ('p',2,2), ('g',0,0), ('g',1,0),
        ('p',3,3), ('g',2,0), ('p',0,5), ('g',0,0)
    ]))

    sink = 0
    iterations = 1000
    for iter in xrange(iterations):
        #checksum = 0
        for test in tests:
            cache = LFUCache(test[0])
            #local = 0
            for op in test[1]:
                if op[0] == 'p':
                    cache.put(op[1], op[2])
                elif op[0] == 'g':
                    r = cache.get(op[1])
            #checksum += local
        #sink = checksum
    # print sink